src/diracx/db/os/utils.py

```diff
@@ -0,0 +1,137 @@
+    # TODO: Make metadata an abstract property
+    fields: dict
+    index_prefix: str
+
+    @abstractmethod
+    def index_name(self, doc_id: int) -> str:
+        ...
+
+    def __init__(self, connection_kwargs: dict[str, Any]) -> None:
+        self._client: AsyncOpenSearch | None = None
+        self._connection_kwargs = connection_kwargs
+
+    @classmethod
+    def available_implementations(cls, db_name: str) -> list[type[BaseOSDB]]:
+        """Return the available implementations of the DB in reverse priority order."""
+        db_classes: list[type[BaseOSDB]] = [
+            entry_point.load()
+            for entry_point in select_from_extension(group="diracx.db.os", name=db_name)
+        ]
+        if not db_classes:
+            raise NotImplementedError(f"Could not find any matches for {db_name=}")
+        return db_classes
+
+    @classmethod
+    def available_urls(cls) -> dict[str, dict[str, Any]]:
+        """Return a dict of available OpenSearch database urls.
+
+        The list of available URLs is determined by environment variables
+        prefixed with ``DIRACX_OS_DB_{DB_NAME}``.
+        """
+        conn_kwargs: dict[str, dict[str, Any]] = {}
+        for entry_point in select_from_extension(group="diracx.db.os"):
+            db_name = entry_point.name
+            var_name = f"DIRACX_OS_DB_{entry_point.name.upper()}"
+            if var_name in os.environ:
+                try:
+                    conn_kwargs[db_name] = json.loads(os.environ[var_name])
+                except Exception:
+                    logger.error("Error loading connection parameters for %s", db_name)
+                    raise
+        return conn_kwargs
+
+    @classmethod
+    def session(cls) -> Self:
+        """This is just a fake method such that the Dependency overwrite has
+        a hash to use"""
+        raise NotImplementedError("This should never be called")
+
+    @property
+    def client(self) -> AsyncOpenSearch:
+        """Just a getter for _client, making sure we entered
+        the context manager"""
+        if self._client is None:
+            raise RuntimeError(f"{self.__class__} was used before entering")
+        return self._client
+
+    @contextlib.asynccontextmanager
+    async def client_context(self) -> AsyncIterator[None]:
+        """Context manage to manage the client lifecycle.
+        This is called when starting fastapi
+
+        """
+        assert self._client is None, "client_context cannot be nested"
+        async with AsyncOpenSearch(**self._connection_kwargs) as self._client:
+            if not await self._client.ping():
+                raise OpenSearchDBUnavailable(
+                    f"Failed to connect to {self.__class__.__qualname__}"
+                )
+            yield
+        self._client = None
+
+    async def __aenter__(self):
+        """This is entered on every request. It does nothing"""
+        assert self._client is None, "client_context hasn't been entered"
```

Comment:
```suggestion
        assert self._client is None, "client_context has not been entered"
```
when it is about printing something, I would avoid any possibly-interpreted-as-special characters.